include/zio/zio_device.h

```diff
@@ -0,0 +1,148 @@
+/*
+ * Copyright (c) 2019 Thomas Burdick <thomas.burdick@gmail.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_ZIO_DEVICE_H_
+#define ZEPHYR_INCLUDE_ZIO_DEVICE_H_
+
+#include <zephyr/types.h>
+#include <device.h>
+
+/**
+ * A ZIO device provides attributes, channels, and triggers for a device.
+ *
+ * Channels are data streams of input or output with attributes themselves which may be
+ * subscribed to.
+ */
+struct zio_channel_desc
+{
+  const char *name;
+  const u16_t kind;
```

Comment:
I'm working on a list of channel types right now, so can propose something here, but reading through the [**iio** API documentation](https://01.org/linuxgraphics/gfx-docs/drm/driver-api/iio/index.html), I do think their approach to nested channel 'type's is useful and flexible.

With iio, when you have multiple instances of the same channel type, you can distinguish them by means of a second (more detailed) type, or via an optional `index` value. See: https://01.org/linuxgraphics/gfx-docs/drm/driver-api/iio/core.html#iio-device-channels

Given that `zio_channel_desc` will almost certainly reside in flash memory, the extra fields shouldn't be a huge burden, but it does make it easier to have two levels of detail about a channel in a single struct.

`type` (used in the iio API) has some naming overlap with the generic concept of data types, but I wonder if 'kind' is an ideal choice here? My gut feeling is that people would be more familiar with the term `type` or maybe `ch_type` if you really want to distinguish it, at the risk of being repetitive?

What is your opinion of the system (described in the link above) of two layers of `kind` or channel `type`?

Reply:
I think that makes some sense, I was thinking that since we have this static array of channel descriptions it really isn't too important necessarily to have a second type/index description field, but perhaps in some scenarios it really is better to do it that way, even if its duplicating in a way the position of the channel description in the static array.

Reply:
We discussed on slack that we likely don't need a secondary enum/int type descriptor, I'm fine renaming this to type